scene-lib/src/scenelib/annotations/io/classfile/ClassAnnotationSceneReader.java

```diff
@@ -4,76 +4,103 @@
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.Type;
+import org.objectweb.asm.TypePath;
+import org.objectweb.asm.TypeReference;
+import scenelib.annotations.Annotation;
+import scenelib.annotations.AnnotationBuilder;
+import scenelib.annotations.AnnotationFactory;
+import scenelib.annotations.Annotations;
+import scenelib.annotations.ArrayBuilder;
+import scenelib.annotations.el.ABlock;
+import scenelib.annotations.el.AClass;
+import scenelib.annotations.el.AElement;
+import scenelib.annotations.el.AField;
+import scenelib.annotations.el.AMethod;
+import scenelib.annotations.el.AScene;
+import scenelib.annotations.el.ATypeElement;
+import scenelib.annotations.el.AnnotationDef;
+import scenelib.annotations.el.BoundLocation;
+import scenelib.annotations.el.LocalLocation;
+import scenelib.annotations.el.RelativeLocation;
+import scenelib.annotations.el.TypeIndexLocation;
+import scenelib.annotations.el.TypePathEntry;
+import scenelib.annotations.field.AnnotationAFT;
+import scenelib.annotations.field.AnnotationFieldType;
+import scenelib.annotations.field.ArrayAFT;
+import scenelib.annotations.field.BasicAFT;
+import scenelib.annotations.field.ClassTokenAFT;
+import scenelib.annotations.field.EnumAFT;
+import scenelib.annotations.field.ScalarAFT;
+import scenelib.annotations.util.coll.VivifyingMap;
 
 /**
- * A <code> ClassAnnotationSceneReader </code> is a
- * {@link org.objectweb.asmx.ClassVisitor} that will insert all annotations it
+ * A {@code ClassAnnotationSceneReader} is a
+ * {@link org.objectweb.asm.ClassVisitor} that will insert all annotations it
  * encounters while visiting a class into a given {@link AScene}.
  *
- * The "read" in <code>ClassAnnotationSceneReader</code> refers to a class
+ * The "read" in {@code ClassAnnotationSceneReader} refers to a class
  * file being read into a scene.  Also see {@link ClassAnnotationSceneWriter}.
  *
  * <p>
  *
  * The proper usage of this class is to construct a
- * <code>ClassAnnotationSceneReader}</code> with an {@link AScene} into which
+ * {@code ClassAnnotationSceneReader} with an {@link AScene} into which
  * annotations should be inserted, then pass this as a
- * {@link org.objectweb.asmx.ClassVisitor} to
- * {@link org.objectweb.asmx.ClassReader#accept}
+ * {@link org.objectweb.asm.ClassVisitor} to
+ * {@link org.objectweb.asm.ClassReader#accept}
  *
  * <p>
  *
  * All other methods are intended to be called only by
- * {@link org.objectweb.asmx.ClassReader#accept},
+ * {@link org.objectweb.asm.ClassReader#accept},
  * and should not be called anywhere else, due to the order in which
- * {@link org.objectweb.asmx.ClassVisitor} methods should be called.
+ * {@link org.objectweb.asm.ClassVisitor} methods should be called.
  */
-public class ClassAnnotationSceneReader
-extends EmptyVisitor {
+public class ClassAnnotationSceneReader extends CodeOffsetAdapter {
   // general strategy:
   // -only "Runtime[In]visible[Type]Annotations" are supported
```

Comment:
I have no idea.  Implies the Runtime[In]VisibleParameterAnnotations are not supported.  But I think they are.  The comment has been there (unchanged) since 2009, perhaps it is no longer correct.